File: src/UI/Web/Blazor/WebApp/App.Client/Services/CustomerReviewService.cs
using BlazorApp.Client.Services.Base;
using ProjectTemplate.Shared.Models;

namespace BlazorApp.Client.Services;

public class CustomerReviewService(IHttpClientFactory httpClientFactory) 
    : BaseService(httpClientFactory, "api/v1/CustomerReview"), ICustomerReviewService
{
    public Task<PagedResponse<CustomerReviewResponseDto>> GetReviewsAsync(int page = 1, int pageSize = 10, string? productName = null, int? minRating = null, bool? isApproved = null, CancellationToken ct = default)
    {
        return GetPagedAsync<CustomerReviewResponseDto>(new { page, pageSize, productName, minRating, isApproved }, ct);
    }

    public Task<CustomerReviewResponseDto?> GetByIdAsync(string id, CancellationToken ct = default)
    {
        return GetAsync<CustomerReviewResponseDto>($"{ResourcePath}/{id}", ct);
    }

    public Task<CustomerReviewResponseDto> CreateAsync(CreateCustomerReviewRequest review, CancellationToken ct = default)
    {
        return PostAsync<CreateCustomerReviewRequest, CustomerReviewResponseDto>(ResourcePath!, review, ct);
    }

    public Task UpdateAsync(string id, UpdateCustomerReviewRequest review, CancellationToken ct = default)
    {
        return PutAsync($"{ResourcePath}/{id}", review, ct);
    }

    public Task ApproveAsync(string id, bool isApproved, CancellationToken ct = default)
    {
        return PatchAsync($"{ResourcePath}/{id}/approve", new ApproveCustomerReviewRequest { IsApproved = isApproved }, ct);
    }

    public Task DeleteAsync(string id, CancellationToken ct = default)
    {
        return DeleteRequestAsync($"{ResourcePath}/{id}", ct);
    }
}

File: src/UI/Web/Angular/src/app/core/services/customer-review.service.ts
import { Injectable } from '@angular/core';
import { BaseService } from './base.service';
import { PagedResponse, CustomerReviewResponse, CreateCustomerReviewRequest, ApproveCustomerReviewRequest } from '../../shared/models';

@Injectable({ providedIn: 'root' })
export class CustomerReviewService extends BaseService<CustomerReviewResponse, CreateCustomerReviewRequest> {
  protected override entityPath = 'api/v1/CustomerReview';

  getReviews(page = 1, pageSize = 10, productName = '', minRating?: number, isApproved?: boolean) {
    let url = `${this.fullUrl}?page=${page}&pageSize=${pageSize}`;
    if (productName) url += `&productName=${encodeURIComponent(productName)}`;
    if (minRating) url += `&minRating=${minRating}`;
    if (isApproved !== undefined) url += `&isApproved=${isApproved}`;
    return this.http.get<PagedResponse<CustomerReviewResponse>>(url);
  }

  approve(id: string, isApproved: boolean) {
    return this.http.patch(`${(this as any).fullUrl}/${id}/approve`, { isApproved } as ApproveCustomerReviewRequest);
  }
}
